satellite/repair/repairer/ec.go

```diff
@@ -0,0 +1,449 @@
+	if err != nil {
+		return nil, err
+	}
+	defer func() { err = errs.Combine(err, downloader.Close()) }()
+
+	pieceBytes, err := ioutil.ReadAll(downloader)
+	if err != nil {
+		return nil, err
+	}
+
+	if int64(len(pieceBytes)) != pieceSize {
+		return nil, Error.New("didn't download the correct amount of data, want %d, got %d", pieceSize, len(pieceBytes))
+	}
+
+	// get signed piece hash and original order limit
+	hash, originalLimit := downloader.GetHashAndLimit()
+	if hash == nil {
+		return nil, Error.New("hash was not sent from storagenode")
+	}
+	if originalLimit == nil {
+		return nil, Error.New("original order limit was not sent from storagenode")
+	}
+
+	// verify order limit from storage node is signed by the satellite
+	if err := verifyOrderLimitSignature(ctx, ec.satelliteSignee, originalLimit); err != nil {
+		return nil, err
+	}
+
+	// verify the hashes from storage node
+	calculatedHash := pkcrypto.SHA256Hash(pieceBytes)
+	if err := verifyPieceHash(ctx, originalLimit, hash, calculatedHash); err != nil {
+		return nil, err
+	}
+
+	return pieceBytes, nil
+}
+
+func verifyPieceHash(ctx context.Context, limit *pb.OrderLimit, hash *pb.PieceHash, expectedHash []byte) (err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	if limit == nil || hash == nil || len(expectedHash) == 0 {
+		return Error.New("invalid arguments")
+	}
+	if limit.PieceId != hash.PieceId {
+		return Error.New("piece id changed")
+	}
+	if !bytes.Equal(hash.Hash, expectedHash) {
+		return Error.New("hashes don't match")
+	}
+
+	if err := signing.VerifyUplinkPieceHashSignature(ctx, limit.UplinkPublicKey, hash); err != nil {
+		return Error.New("invalid piece hash signature")
+	}
+
+	return nil
+}
+
+func verifyOrderLimitSignature(ctx context.Context, satellite signing.Signee, limit *pb.OrderLimit) (err error) {
+	if err := signing.VerifyOrderLimitSignature(ctx, satellite, limit); err != nil {
+		return Error.New("invalid order limit signature: %v", err)
+	}
+
+	return nil
+}
+
+// Repair takes a provided segment, encodes it with the provided redundancy strategy,
+// and uploads the pieces in need of repair to new nodes provided by order limits.
+func (ec *ECRepairer) Repair(ctx context.Context, limits []*pb.AddressedOrderLimit, privateKey storj.PiecePrivateKey, rs eestream.RedundancyStrategy, data io.Reader, expiration time.Time, timeout time.Duration, path storj.Path) (successfulNodes []*pb.Node, successfulHashes []*pb.PieceHash, err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	pieceCount := len(limits)
+	if pieceCount != rs.TotalCount() {
+		return nil, nil, Error.New("size of limits slice (%d) does not match total count (%d) of erasure scheme", pieceCount, rs.TotalCount())
+	}
+
+	if !unique(limits) {
+		return nil, nil, Error.New("duplicated nodes are not allowed")
+	}
+
+	readers, err := eestream.EncodeReader(ctx, ec.log, ioutil.NopCloser(data), rs)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	type info struct {
```

Comment:
I think this is a carry over from the ec Repair function, but I think a more descriptive name for this struct and the `infos` channel could make this part easier to understand.

Reply:
Hmm I am not sure what a better name for the struct and channel would be, but maybe adding a comment could help?

Reply:
Added some comments describing the purpose of the struct + channel

Reply:
thank you, I find those helpful!